Feature/theme and theme browser (#370)
pxblue/doc-it: src/docs/style/themes.mdx

import { ImageGrid, DemoCard, MaterialDesignDescription, Divider, ThemeBrowser } from '../../app/components';
import { REGULAR_WIDTH_STYLE } from '../../app/shared';
import LightThemeColors from '../../app/assets/styles/themes/light-theme-colors.png';

<!-- KEYWORDS: design style theme light dark mode -->

# Themes

> One of the key benefits of PX Blue is consistency of look & feel between different applications. This is achieved primarily through the use of PX Blue themes. PX Blue offers two themes: a blue theme (light theme) and a blue dark theme.

Reviewer: ```suggestion
One of the key benefits of PX Blue is consistency of look & feel between different applications. This is achieved primarily through the use of PX Blue themes. PX Blue offers two themes: a light theme (blue) and a dark theme (blueDark).
```

        style={{ marginRight: 0 }}
    />
</div>

## Explore Our Themes

> Have a look at the component theme explorer below and learn how PX Blue themes work on a high level. The theme explorer selectively showcases some of the most representative components from Material Design and from our component library that fabrics our theme.

Reviewer: ```suggestion
Have a look at the component theme explorer below and learn how PX Blue themes work at a high level. Here we showcase some of the most representative components from Material Design and how our themes affect them.
```

<ThemeBrowser />

> You are also encouraged to visit our code examples, modify our code, and see the magic happens in real time.

Reviewer: ```suggestion
You are also encouraged to explore our more detailed theme examples where you can modify the code and see the effects in real time.
```

<React.Fragment>
    <DemoCard
        repository={'showcase-demo'}
        angular={`https://stackblitz.com/github/pxblue/angular-showcase-demo/tree/master`}
        react={`https://codesandbox.io/s/github/pxblue/react-showcase-demo/tree/master`}
    />
    <Divider />
</React.Fragment>

# Blue Theme / Light Theme

> A SCREENSHOT HERE

> The blue theme, inherited and expanded from Material Design's baseline theme, is the default theme for all PX Blue applications. With a clean color vibrance and high text legibility, it is designed to suit day to day use on a computer monitor, on-site facility inspection with a tablet, and on a mobile phone screen. It should be applied at the root level of your app.

Reviewer: Technically all of our themes are based on the MD baseline theme...I don't think we need to explicitly say that.

Author: I've moved it to the top. 

The term "theme" technically consists of colors, typographies, shapes, icons, etc. I find it worth mentioning that our theme is no different from their baseline theme except for colors and typographies (design patterns are not part of a theme), as part of an answer to the question "what's the diff between MD and PXB"

> The blue theme uses our [iconic Blue 500 color](/style/color) (`#007bc1`) throughout the application, notably in toolbars, buttons and icons. Occasionally you will notice a accent / secondary color, Light Blue 500, being applied to interactive UI elements such as form controls. Blue theme's base background color is White 200, although the surface elements are often shaded in pure White (White 50).

Reviewer: ```suggestion
The light theme uses our iconic [Blue 500](/style/color) color (`#007bc1`) throughout the application, notably in toolbars, buttons and icons. In some places, Light Blue 500 is applied as an accent color on interactive UI elements such as form controls. The light theme's base background color is White 200, although the surface elements are often shaded in White 50 (pure white).
```

<ImageGrid images={[LightThemeColors]} regularWidth />
<Divider />

> # Blue Dark Theme

Reviewer: ```suggestion
# Dark Theme / Blue Dark Theme
```

> A SCREENSHOT HERE

> The blue dark theme is constructed in a similiar fashion just like the blue theme, but is intended primarily for applications used in low-light conditions. As the name indicates, the blue dark theme is dominated by different shades of dark colors, and uses colors only to for interactive components and status displays.

Reviewer: ```suggestion
The dark theme is intended primarily for applications used in low-light conditions. As the name indicates, it utilizes different shades of dark colors (grays and blacks), and uses other colors only for interactive components and status displays.
```

> Some fancy color stuff

> To accomodate for [accessibility requirements](/design/accessibility), PX Blue uses 200 colors when a text shows by itself on a dark background, such as in an outlined button and a text field label. For a "contained" element, such as in an elevated / contained button, the style remains Blue 500 on White.

Reviewer: ```suggestion
To accommodate for [accessibility requirements](/design/accessibility), PX Blue uses the 200 shade of colors when text is used on a dark background, such as in an outlined button and a text field label. For "contained" elements, such as in an elevated / contained button, the same style is used as in the light theme (500 shades).
```

> A PIC SHOWING
>
> -   YEA: BLUE200 on dark
> -   YEA: WHITE50 on BLUE500
> -   NAY: DARK on BLUE200

> A PIC showing that carefully using transparency for bg colors make your stuff looks good on light and dark theme, such as the list item tag

<Divider />

# Developers

## Angular and Ionic

> PX Blue offers the theming support via the package [@pxblue/angular-themes](https://www.npmjs.com/package/@pxblue/angular-themes), which is a collection of SCSS stylesheets that are imported into your application and applied through classes (`pxb-blue` or `pxb-dark`), to both PX Blue components and Angular Material components.

Reviewer: We don't need to link to the same thing twice here
```suggestion
Angular themes are available from our @pxblue/angular-themes package, which is a collection of SCSS stylesheets that are imported into your application and applied via classes (`pxb-blue` or `pxb-dark`) to both PX Blue components and Angular Material components.
```

-   @pxblue/angular-themes: [NPM](https://www.npmjs.com/package/@pxblue/angular-themes), [GitHub](https://github.com/pxblue/themes/tree/master/angular)
-   [Theming Guide](https://material.angular.io/guide/theming) on Angular Material

If you are using Ionic for hybrid application development, you can utilize our Angular themes as well. See [our Ionic getting started guide](https://pxblue.github.io/development/frameworks-mobile/ionic).

## React

> PX Blue's React themes are packaged through [@pxblue/react-themes](https://www.npmjs.com/package/@pxblue/react-themes). It contains JSON objects that are consumed by the theme mechanism for Material UI components (`<ThemeProvider />`). The package also contains some specific rules defined in [JSS](https://material-ui.com/customization/css-in-js/).

Reviewer: ```suggestion
React themes are available from our @pxblue/react-themes package, which contains JSON objects that are consumed by the theme mechanism for Material UI components (`<ThemeProvider />`), including some specific styling rules defined in [JSS](https://material-ui.com/customization/css-in-js/).
```

-   @pxblue/react-themes: [NPM](https://www.npmjs.com/package/@pxblue/react-themes), [GitHub](https://github.com/pxblue/themes/tree/master/react)
-   [Theming Guide](https://material-ui.com/customization/themes/) on Material-UI

## React Native

> React Native themes are encapsulated in [@pxblue/react-native-themes](https://www.npmjs.com/package/@pxblue/react-native-themes), which relies on React Native Paper's theming mechanism.

Reviewer: ```suggestion
React Native themes are available from our @pxblue/react-themes package, which utilize React Native Paper's theming mechanism. This package contains two separate dark themes which must be applied to different components to achieve the correct appearance (see the package readme for details).
```

-   @pxblue/react-native-themes: [NPM](https://www.npmjs.com/package/@pxblue/react-native-themes), [GitHub](https://github.com/pxblue/themes/tree/master/react-native)
-   [Theming Guide](https://callstack.github.io/react-native-paper/theming.html) on React Native Paper

## Color Palette

> All of the [PX Blue colors](/style/color) are available individually ([@pxblue/colors](https://www.npmjs.com/package/@pxblue/colors)) for use in your application. However, for the most part, you should rely on our themes packages to alter the styles.

Reviewer: ```suggestion
All of the [PX Blue colors](/style/color) are available individually ([@pxblue/colors](https://www.npmjs.com/package/@pxblue/colors)) for use in your application. However, in most cases, you should use values from the themes to alter styles for better future-proofing.
```

The blue theme is the default theme for all PX Blue applications. With a clean color vibrance and high text legibility, it is designed to suit day-to-day use on a computer monitor, on-site facility inspection with a tablet, and on a mobile phone screen. It should be applied at the root level of your app.

The light theme uses our iconic [Blue 500](/style/color) color (`#007bc1`) throughout the application, notably in toolbars, buttons, and icons. In some places, Light Blue 500 is applied as an accent/secondary color on interactive UI elements such as form controls. The light theme's base background color is White 200, although the surface elements are often shaded in White 50 (pure white).

<ImageGrid images={[LightThemeColors]} regularWidth />
<Divider />

# Dark Theme / Blue Dark Theme

<div style={REGULAR_WIDTH_STYLE}>
    <ImageGrid images={[DarkThemeExample]} regularWidth caption={'An example of a dark-themed application.'} />
</div>

The dark theme is intended primarily for applications used in low-light conditions. As the name indicates, it utilizes different shades of dark colors (grays and blacks) and uses other colors only for interactive components and status displays.

> <ImageGrid images={[DarkThemeColors]} regularWidth />

Author: I can do that. I feel like since it is explained by the paragraph immediately following the image, I kinda skipped it here.

> <ImageGrid images={[DarkThemeColors]} regularWidth />

Reviewer: Yeah, I know what you mean. It just catches the reader (me) by surprise to see the color on the left and then a different color on the right. Maybe we could integrate some error text in the example on the right in both light/dark and refer to that as the error color instead of the badge (which isn't actually necessarily related to warnings if you think about it).

To accommodate for [accessibility requirements](/design/accessibility), PX Blue uses the 200 shade of colors when text is used on a dark background, such as in an outlined button and a text field label. For "contained" elements, such as in an elevated / contained button, the same style is used as in the light theme (500 shades).

<div style={REGULAR_WIDTH_STYLE}>
    <ImageGrid

>         images={[OutlinedAndFlat]}

Author: I don't think docit has any images with green _bars_ in them — the idea was so that people with color blindness can single them out better... Although I guess there are better ways to handle that.

There certainly are images with green _backgrounds_. However, a green background would not work here, as the dark background is critical in illustrating my point.

>         images={[OutlinedAndFlat]}

Reviewer: Ah, okay, I think I must be misremembering. I thought we used red and green bars, but if we're not, then ignore me.

        caption={`You may use the 200 shade of text colors on a dark background (#1) or White 50 on 500 shades (#2). However, the 200 shades should never be used as a background color (#3), as they glare in a dark environment.`}
        regularWidth
    />
    <ImageGrid
        images={[DarkThemeElevation]}
        caption={[
            'In dark theme, you can express surface elevation with:',
            '1: A different surface color. (Optionally, you may apply a drop shadow to further define the edge.)',
            '2: Use outlined component style.',

>             '3: However, you should never use SOLELY the drop shadow, no matter how dark the shadow is.',

Reviewer: ```suggestion
            '3: However, you should never rely SOLELY on the drop shadow, no matter how dark it is.',
```
